junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/ForkJoinPoolHierarchicalTestExecutorService.java

```diff
@@ -112,38 +122,30 @@ private static Callable<ForkJoinPool> sinceJava7ConstructorInvocation(ParallelEx
 		return () -> new ForkJoinPool(configuration.getParallelism(), threadFactory, null, false);
 	}
 
-	@Override
-	public Future<Void> submit(TestTask testTask) {
-		ExclusiveTask exclusiveTask = new ExclusiveTask(testTask);
-		if (!isAlreadyRunningInForkJoinPool()) {
-			// ensure we're running inside the ForkJoinPool so we
-			// can use ForkJoinTask API in invokeAll etc.
-			return forkJoinPool.submit(exclusiveTask);
+	@SuppressWarnings("try")
+	private static void executeWithLock(TestTask testTask) {
+		ResourceLock resourceLock = testTask.getResourceLock();
+		if (!(Thread.currentThread() instanceof ForkJoinWorkerThread)) {
+			resourceLock.release();
+		}
```

Comment:
The `canRunTestsIsolatedFromEachOtherWithNestedCases` test from `FixedThreadsParallelExecutionIntegrationTests` was stuck in a deadlock until I added these lines. I spent some time debugging and it seems to me that all of the `org.junit.platform.engine.support.hierarchical.SingleLock` instances created for the tasks are somehow locked by a fork-join thread before the tasks are scheduled in the fixed thread pool, but no matter how many breakpoints I added, I couldn't figure out where that's happening.